### PR TITLE
New style vstack

### PR DESCRIPTION
**Pull request by hvy**
New style vstack to support double backprop.

### CODE REVIEW

**Comment by hvy** [comment]
Rebased on top of master.

File: chainer/functions/array/vstack.py [hunk 1]
```diff
@@ -26,26 +27,24 @@ def check_type_forward(self, in_types):
...
+        ndim = len(self.inputs[0].shape)
+        if len(self.inputs) == 1:
+            if ndim <= 1:
+                return gy.reshape(self.inputs[0].shape),
+            return gy,
 
-        xp = cuda.get_array_module(*gy)
+        if ndim <= 1:
+            gx = chainer.functions.split_axis(gy, len(self.inputs), 0)
+            return [g.reshape(self.inputs[0].shape) for g in gx]
```
Comment (takagi) [inline chainer/functions/array/vstack.py]: `gx` and `gxs` would be better than `g` and `gx`?

File: chainer/functions/array/vstack.py [hunk 2]
```diff
@@ -26,26 +27,24 @@ def check_type_forward(self, in_types):
                 type_check.expect(in_types[0].shape[d] == in_types[i].shape[d])
 
     def forward(self, xs):
-        self.retain_inputs(())
         self._in_shapes = [x.shape for x in xs]
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/array/vstack.py]: Removed the unused variable.

File: chainer/functions/array/vstack.py (hunk 1 above)
Comment (hvy) [reply chainer/functions/array/vstack.py]: Renamed variables accordingly.

**Comment by hvy** [comment]
@takagi san, thank you for the review! I made some changes to the code according to your feedback.

**Pull request merged**